src/aws/flb_s3_local_buffer.c

```diff
@@ -414,3 +414,110 @@ int flb_remove_chunk_files(struct flb_local_chunk *c)
     }
     return ret;
 }
+
+/* Generic replace function for strings. */
+char* replace_uri_tokens(const char* original_string, const char* current_word, const char* new_word) 
+{ 
+    char *result; 
+    int i = 0;
+    int count = 0; 
+    int new_word_len = strlen(new_word); 
+    int old_word_len = strlen(current_word); 
+  
+    for (i = 0; original_string[i] != '\0'; i++) { 
+        if (strstr(&original_string[i], current_word) == &original_string[i]) { 
+            count++; 
+            i += old_word_len - 1; 
+        } 
+    }
+
+    result = flb_sds_create_size(i + count * (new_word_len - old_word_len) + 1);
+    if (!result) {
+        flb_errno();
+        return NULL;
+    }
+
+    i = 0; 
+    while (*original_string) { 
+        if (strstr(original_string, current_word) == original_string) { 
+            strcpy(&result[i], new_word); 
+            i += new_word_len; 
+            original_string += old_word_len; 
+        } 
+        else
+            result[i++] = *original_string++; 
+    } 
+
+    result[i] = '\0'; 
+    return result; 
+} 
+  
+//* Constructs S3 object key as per the format. */
+flb_sds_t get_s3_key(const char *format, time_t time, char *tag, char *tag_delimiter)
+{
+    int i = 0;
+    int max_key_len = (strlen(format) + strlen(tag) + 1);
+    char tag_idx[9];
+    char open[2] = "[";
+    char close[2] = "]";
+    char *tag_token = NULL;
+    char tag_descriptor[10] = "$TAG";
+    char datestamp[max_key_len];
+    flb_sds_t s3_key = NULL;
+    flb_sds_t tmp_key = NULL;
+    char *tmp_tag = NULL;
+    struct tm *gmt = NULL;
+    
+    tmp_tag = flb_malloc(strlen(tag) + 1);
+    s3_key = flb_sds_create_size(max_key_len);
+    if (!s3_key) {
+        flb_errno();
+        return NULL;
+    }
+    strcpy(s3_key, format);
+    strcpy(tmp_tag, tag);
+
+    /* Finds all occurences of $TAG[*] and replaces then with the right token from tag.*/
+    tag_token = strtok(tmp_tag, tag_delimiter);
+    
+    while(tag_token != NULL) {
+        char tmp[20] = "";
+        sprintf(tag_idx, "%d", i);
+        strcat(tmp, tag_descriptor);
+        strcat(tmp, open);
+        strcat(tmp, tag_idx);
+        strcat(tmp, close);
+
+        tmp_key = replace_uri_tokens(s3_key, tmp, tag_token);
+        if (!tmp_key) {
+            flb_errno();
+            return NULL;
+        }
+        flb_sds_destroy(s3_key);
+        s3_key = tmp_key;
+        tag_token = strtok(NULL, tag_delimiter);
+        i++;
+    }
+ 
+    /* Finds all occurences of $TAG and replaces then with the entire tag. */
+    tmp_key = replace_uri_tokens(s3_key, tag_descriptor, tag);
+    if (!tmp_key) {
+        flb_errno();
+        return NULL;
+    }
+    flb_sds_destroy(s3_key);
+    s3_key = tmp_key; 
+ 
+    gmt = gmtime(&time);
+    strftime(datestamp, sizeof(datestamp) - 1, s3_key, gmt);
+    tmp_key = replace_uri_tokens(s3_key, s3_key, datestamp); 
```

Comment:
I don't understand this line... the call to strftime will fill `datestamp` with what you want... don't see why you need the call to replace_uri_tokens here? After the strftime call you should be done.

Reply:
Also now I see how max_key_len is used... its the length of the final string, which is only used here (datestamp is a bad var name btw- since this is the final string it should be something else... or just call it tmp and re-assign it to s3_key afterwords...).

I think it would be better if you used flb_sds_create_size here to alloc a string which is as long as is needed. At this point, you have constructed s3_key with all the tag parts, the only thing left is adding the date parts.

The big question is- how much could the length of the string increase by setting the timestamp?

Look at the docs for strftime: https://man7.org/linux/man-pages/man3/strftime.3.html

From my reading, some of the format specifiers can take a 2 char input string and output a 10 char output string. So the final length could be 10 * strlen(s3_key). (Remember that the user input could be anything).

Another idea is that from the docs, strftime apparently returns 0 if there is insufficient memory available. You could run it in a loop where each time you alloc a longer string until it succeeds.

Reply:
The reason I called replace_uri_tokens here is because datestamp is not the expected return format. I can create a tmp with flb_sds_create_size and then re-assign that to s3_key.